accounts/service.go

```diff
@@ -0,0 +1,628 @@
+func (s *InterceptorService) Accounts() ([]*OffChainBalanceAccount, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.store.Accounts()
+}
+
+// RemoveAccount finds an account by its ID and removes it from the DB.
+func (s *InterceptorService) RemoveAccount(id AccountID) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// Are we currently tracking any payments?
+	for hash, payment := range s.pendingPayments {
+		if payment.accountID != id {
+			continue
+		}
+
+		// Let's remove the payment (which also cancels the tracking).
+		err := s.removePayment(hash, lnrpc.Payment_FAILED)
+		if err != nil {
+			return err
+		}
+	}
+
+	return s.store.RemoveAccount(id)
+}
+
+// CheckBalance ensures an account is valid and has a balance equal to or larger
+// than the amount that is required.
+func (s *InterceptorService) CheckBalance(id AccountID,
+	requiredBalance lnwire.MilliSatoshi) error {
+
+	s.RLock()
+	defer s.RUnlock()
+
+	// Check that the account exists, it hasn't expired and has sufficient
+	// balance.
+	account, err := s.store.Account(id)
+	if err != nil {
+		return err
+	}
+
+	if account.HasExpired() {
+		return ErrAccExpired
+	}
+
+	var inFlightAmt int64
+	for _, pendingPayment := range s.pendingPayments {
+		inFlightAmt += int64(pendingPayment.fullAmount)
+	}
+
+	availableAmount := account.CurrentBalance - inFlightAmt
+	if availableAmount < int64(requiredBalance) {
+		return ErrAccBalanceInsufficient
+	}
+
+	return nil
+}
+
+// AssociateInvoice associates a generated invoice with the given account,
+// making it possible for the account to be credited in case the invoice is
+// paid.
+func (s *InterceptorService) AssociateInvoice(id AccountID,
+	hash lntypes.Hash) error {
+
+	s.Lock()
+	defer s.Unlock()
+
+	account, err := s.store.Account(id)
+	if err != nil {
+		return err
+	}
+
+	account.Invoices[hash] = struct{}{}
+	s.invoiceToAccount[hash] = id
+	return s.store.UpdateAccount(account)
+}
+
+// invoiceUpdate credits the account an invoice was registered with, in case the
+// invoice was settled.
+func (s *InterceptorService) invoiceUpdate(invoice *lndclient.Invoice) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// We update our indexes each time we get a new invoice from our
+	// subscription. This might be a bit inefficient but makes sure we don't
+	// miss an update.
+	needUpdate := false
+	if invoice.AddIndex > s.currentAddIndex {
+		s.currentAddIndex = invoice.AddIndex
+		needUpdate = true
+	}
+	if invoice.SettleIndex > s.currentSettleIndex {
+		s.currentSettleIndex = invoice.SettleIndex
+		needUpdate = true
+	}
+
+	if needUpdate {
+		err := s.store.StoreLastIndexes(
+			s.currentAddIndex, s.currentSettleIndex,
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	// The invoice hasn't been settled yet, there is nothing for us to do.
+	// If it eventually settles, we'll be called again.
+	if invoice.State != channeldb.ContractSettled {
+		return nil
+	}
+
+	// The invoice was settled, let's now debit the account. But only if the
```

Comment:
```suggestion
	// The invoice was settled, let's now credit the account. But only if the
```

Reply:
I think this is correct: ["A debit entry in an account represents a transfer of value to that account, and a credit entry represents a transfer from the account."](https://en.wikipedia.org/wiki/Debits_and_credits).

Reply:
> I think this is correct: ["A debit entry in an account represents a transfer of value to that account, and a credit entry represents a transfer from the account."](https://en.wikipedia.org/wiki/Debits_and_credits).

Wow, that goes against my understanding of the words but apparently in the accounting world (I am not an accountant) "debit" does indeed mean adding value to an account and "credit" means subtracting value. But that seems to go against the "common" meaning of the words, according to [this article](https://www.sapling.com/4827482/debit-vs-credit-accounting):

> Outside of the accounting world, the term "debit" usually refers to money removed from a consumer bank account, such as money removed from your checking account when you buy groceries. Similarly, "credit" usually refers to money added to a consumer bank account account, or to money that is otherwise freely yours to spend

The dictionary definitions of the words also are in line with debit = payment/subtraction and credit = received/addition.

My 2 cents is that the common definitions of the words would be preferable, since this will be read by software developers and not accountants. It also seems to me that most existing usage of credit/debit is already in line with the common definitions, e.g.: https://github.com/lightninglabs/lightning-terminal/pull/363/files#diff-3bfb289868a7cc73be6cb96f5f75cf79f1b8f14395e62ffa03283ffb2aa07e6cR219

Reply:
Hmm, no wonder I always get those confused then, if even Wikipedia disagrees with most use cases. And yes, after taking a closer look, I did use the terms mostly in the non-Wikipedia way so I fixed the two occurrences you highlighted.